### DISCUSSION TITLE
Fetching The Device State

### DISCUSSION THREAD

**Discussion by seflless** [djipco/webmidi#231] [Q&A]
I'm looking for a way to fetch the state of each knob/button/touchpad on my MIDI keyboard. I have made a demo where I made a digital version of my Korg nanoKEY Studio keyboard. It can respond to and update the visualization of any changed knob/button/touchpad states. Without a way to fetch the initial state, my knobs, and stateful buttons are not correct until I get a change event and can at least know that element's state. 

https://user-images.githubusercontent.com/1845/151546483-86de0671-e430-4f65-89d5-3b6a4fcebed1.mov

I discussed this demo a bit [here on Twitter too.](https://twitter.com/seflless/status/1486327881500344331?s=20
)

### Discussion on Forum
There was an [answer on the other forum](https://webmidijs.org/forum/discussion/28/requesting-current-values-of-control-parameters-from-a-device.html) that was related to the same goals, but I can't figure out how to use it. Basically this was significant part of it. You call `output.sendSysex(0x42, [0x30, 0x7e, 0x10]);`.  There's also some details about where to find the device/manufacturer and etc IDs.

What event do I have to listen to get the device state that is sent back?

**Comment by djipco** [comment]
> What event do I have to listen to get the device state that is sent back?

You should listen for the `sysex` message that the machine will send back after you sent the request.

**Comment by seflless** [comment]
Ah, it was hard to find a list of events there are without spelunking in code. I don't ever get a message though, now that I added a listener:

### Sending
```
// Outputs
  WebMidi.outputs.forEach((output) => {
    output.sendSysex(0x42, [0x30, 0x7e, 0x10]);
```

### Listening
```
output.addEventListener("sysex", sysexMessageReceived);
```

**Reply by djipco** [reply]
You need to add the listener to the `Input` and not the `Output`. You use the `Output` to send the request and the `Input` to receive the answer (which are two completely separate things). 

You can view the full list of events by checking out the documentation for the [`Input`](https://webmidijs.org/api/classes/Input) and [`Output`](https://webmidijs.org/api/classes/Output) classes. In particular, check out the documentation for the `addListener()` method.

**Comment by seflless** [comment]
Ok, good catch. I fixed it so I'm adding sysex listener on the input device. I don't unfortunately get a sysex message ever.

### My sysex ping
```
WebMidi.enable({ sysex: true })
  .then(onEnabled)
  .catch((err) => alert(err));

window.WebMidi = WebMidi;

function onEnabled() {
  WebMidi.outputs.forEach((output) => {
    output.sendSysex(0x42, [0x30, 0x7e, 0x10]);
    console.log("outputs", output.manufacturer, output.name);
  });
}
```

### My Listening Registration
```
function startListening() {
  for (const input of midiIn) {
    input.addEventListener("midimessage", midiMessageReceived);
    input.addEventListener("sysex", sysexMessageReceived);
  }
}

function sysexMessageReceived(event) {
  console.log(event);
}
```

Perhaps I need different numbers sent for my Korg nanoKEY Studio device? Sorry for the dumb questions!

**Reply by djipco** [reply]
I don't think you have the correct message format. When I look at the [Korg sysex implementation](https://ia803205.us.archive.org/21/items/micro-korg-sysex-implementation/MicroKorg%20SYSEX%20implementation.txt), this is the format to dump the current program's data:

```
(6) CURRENT PROGRAM DATA DUMP                                    R/T
+----------------+--------------------------------------------------+
|     Byte       |             Description                          |
+----------------+--------------------------------------------------+
| F0,42,3g,58    | EXCLUSIVE HEADER                                 |
| 0100 0000 (40) | CURRENT PROGRAM DATA DUMP              40H       |
| 0ddd dddd (dd) | Data                                  (NOTE 1,5) |
|     :          |  :                                               |
| 1111 0111 (F7) | EOX                                              |
+----------------+--------------------------------------------------+
```
I have no clue what `3g` means. There is no `g` in hexadecimal notation so it must stand for something else.
